src/platform/i386/chal_pgtbl.c

```diff
@@ -0,0 +1,188 @@
+#include "shared/cos_types.h"
+#include "captbl.h"
+#include "pgtbl.h"
+#include "cap_ops.h"
+#include "liveness_tbl.h"
+#include "retype_tbl.h"
+
+int
+chal_pgtbl_kmem_act(pgtbl_t pt, u32_t addr, unsigned long *kern_addr, unsigned long **pte_ret)
+{
+	struct ert_intern *pte;
+	u32_t              orig_v, new_v, accum = 0;
+
+	assert(pt);
+	assert((PGTBL_FLAG_MASK & addr) == 0);
+
+	/* get the pte */
+	pte = (struct ert_intern *)__pgtbl_lkupan((pgtbl_t)((u32_t)pt | PGTBL_PRESENT), addr >> PGTBL_PAGEIDX_SHIFT,
+	                                          PGTBL_DEPTH, &accum);
+	if (unlikely(!pte)) return -ENOENT;
+	if (unlikely(__pgtbl_isnull(pte, 0, 0))) return -ENOENT;
+
+	orig_v = (u32_t)(pte->next);
+	if (unlikely(!(orig_v & PGTBL_COSFRAME))) return -EINVAL; /* can't activate non-frames */
+	if (unlikely(orig_v & PGTBL_COSKMEM)) return -EEXIST;     /* can't re-activate kmem frames */
+	assert(!(orig_v & PGTBL_QUIESCENCE));
+
+	*kern_addr = (unsigned long)chal_pa2va((paddr_t)(orig_v & PGTBL_FRAME_MASK));
+	new_v      = orig_v | PGTBL_COSKMEM;
+
+	/* pa2va (value in *kern_addr) will return NULL if the page is
+	 * not kernel accessible */
+	if (unlikely(!*kern_addr)) return -EINVAL; /* cannot retype a non-kernel accessible page */
+	if (unlikely(retypetbl_kern_ref((void *)(new_v & PGTBL_FRAME_MASK)))) return -EFAULT;
+
+	/* We keep the cos_frame entry, but mark it as COSKMEM so that
+	 * we won't use it for other kernel objects. */
+	if (unlikely(cos_cas((unsigned long *)pte, orig_v, new_v) != CAS_SUCCESS)) {
+		/* restore the ref cnt. */
+		retypetbl_deref((void *)(orig_v & PGTBL_FRAME_MASK));
+		return -ECASFAIL;
+	}
+	/* Return the pte ptr, so that we can release the page if the
+	 * kobj activation failed later. */
+	*pte_ret = (unsigned long *)pte;
```

Comment:
does the code that calls this not assume a specific layout for the `pte`?  If so, should that code also be in the chal?

Reply:
Yes, it should also be in chal.